docs/design/features/NonGC-Heap.md

```diff
@@ -0,0 +1,125 @@
+# NonGC Heap
+
+.NET 8.0 introduces a new concept called **NonGC Heap** - a specialized heap that is not managed by the Garbage Collector (GC) and is designed to store immortal objects with certain benefits for GC and code generation (codegen). Although the name is new, the feature is based on pre-existing **Frozen Segments** which were added long time ago to serve a similar purpose and were used on CoreRT/NativeAOT. What's changed in .NET 8.0 is that this functionality has been exposed to users through public profiling and debugging APIs, and it is now heavily leveraged for several codegen optimizations in RyuJIT. A general overview of the .NET managed heap can be visualized as shown in the following mermaid diagram:
+
+```mermaid
+flowchart
+    heap(.NET Managed Heap) --> gcheap
+    heap --> nongcheap
+    gcheap --> soh
+    gcheap --> poh
+    gcheap --> loh
+
+    gcheap("GC Heap")
+    soh("Small Object Heap\n(SOH)")
+    poh("Pinned Object Heap\n(SOH)")
+    loh("Large Object Heap\n(SOH)")
+    nongcheap("Non-GC Heap\n(Not managed by GC)")
+
+    style nongcheap stroke:green
+```
+
+The basic idea that certain kinds of objects are essentially immortal and will never be collected, hence, we can put them into a separate storage where they are never scanned or compacted. This has significant implications for RyuJIT, as it can directly embed GC references (or even byrefs) in codegen. For example, in order to access a string object that represents a string literal `"Hello"` (all string literals are interned and therefore immortal) an indirect load is required to access that string. In pre-.NET 8.0 runtimes, the following C# code:
+```cs
+string Test() => "Hello";
+```
+produces the following codegen:
+```asm
+; Method Program:Test():System.String:this
+       mov      rax, 0xAABBCCDDEEFF
+       mov      rax, gword ptr [rax]
+       ret
+```
+Here, VM must create a small pinned handle that keeps a reference to the source string object that is kept in a normal heap. The address of that pinned handle (`0xAABBCCDDEEFF` in the example above) is them passed to RyuJIT, enabling it to access the original object via an indirect load and to keep the codegen GC-friendly in case if the source string object is moved to a different location during a sudden heap compaction. The pinned handle must be rooted to ensure it remains alive. By pinning the object itself instead, JIT can "bake" the direct reference, resulting in a more efficient codegen:
```

Comment:
```suggestion
Here, VM must create a small pinned handle that keeps a reference to the source string object that is kept in a normal heap. The address of that pinned handle (`0xAABBCCDDEEFF` in the example above) is then passed to RyuJIT, enabling it to access the original object via an indirect load and to keep the codegen GC-friendly in case if the source string object is moved to a different location during a sudden heap compaction. The pinned handle must be rooted to ensure it remains alive. By pinning the object itself instead, JIT can "bake" the direct reference, resulting in a more efficient codegen:
```